File: tmp/see_agent/sandbox/collector.py
"""Sandbox violation collector — reads macOS unified log for deny events."""

from __future__ import annotations

import asyncio
import json
import logging
import re

logger = logging.getLogger(__name__)

_PATH_RE = re.compile(r'path\s+"([^"]+)"', re.IGNORECASE)
_OP_RE = re.compile(r"deny\(1\)\s+([\w-]+)", re.IGNORECASE)


class SandboxViolationCollector:
    """Collect sandbox deny records from macOS unified log."""

    async def collect(
        self,
        agent_pid: int,
        since_minutes: int = 60,
    ) -> list[dict[str, str]]:
        """Extract sandbox violations for a given PID."""
        try:
            proc = await asyncio.create_subprocess_exec(
                "log", "show",
                "--predicate",
                f'processID == {agent_pid} AND category == "Sandbox"',
                "--last", f"{since_minutes}m",
                "--style", "json",
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
            )
            stdout, _ = await proc.communicate()
            return self._parse_violations(stdout.decode())
        except FileNotFoundError:
            logger.warning("'log' command not found — cannot collect violations")
            return []
        except Exception:
            logger.exception("Failed to collect sandbox violations")
            return []

    def _parse_violations(self, log_output: str) -> list[dict[str, str]]:
        if not log_output.strip():
            return []
        try:
            entries = json.loads(log_output)
        except json.JSONDecodeError:
            return []

        violations: list[dict[str, str]] = []
        for entry in entries:
            msg = entry.get("eventMessage", "")
            if "deny" not in msg.lower():
                continue
            path = self._extract_path(msg)
            operation = self._extract_operation(msg)
            violations.append({
                "timestamp": entry.get("timestamp", ""),
                "operation": operation,
                "path": path,
            })
        return violations

    def _extract_path(self, msg: str) -> str:
        m = _PATH_RE.search(msg)
        return m.group(1) if m else ""

    def _extract_operation(self, msg: str) -> str:
        m = _OP_RE.search(msg)
        return m.group(1) if m else "unknown"

File: see/src/sandbox/collector.rs
use serde::Serialize;

/// A single sandbox violation event from macOS unified logs.
#[derive(Debug, Clone, Serialize)]
pub struct SandboxViolation {
    pub timestamp: String,
    pub operation: String,
    pub path: String,
}

/// Collect sandbox deny events for a given process ID from macOS unified logs.
///
/// Uses `log show` to query sandbox category events for the given PID
/// within the last `minutes` minutes.
pub async fn collect_violations(pid: u32, minutes: u32) -> Vec<SandboxViolation> {
    let predicate = format!(
        "processID == {pid} AND category == \"Sandbox\""
    );
    let last = format!("{minutes}m");

    let output = match tokio::process::Command::new("log")
        .args(["show", "--predicate", &predicate, "--last", &last, "--style", "json"])
        .output()
        .await
    {
        Ok(o) => o,
        Err(_) => return Vec::new(),
    };

    if !output.status.success() {
        return Vec::new();
    }

    let text = String::from_utf8_lossy(&output.stdout);
    parse_log_output(&text)
}

/// Parse JSON log output from `log show --style json` and extract sandbox deny events.
fn parse_log_output(text: &str) -> Vec<SandboxViolation> {
    let mut violations = Vec::new();

    // The output is a JSON array of log entries
    let entries: Vec<serde_json::Value> = match serde_json::from_str(text) {
        Ok(v) => v,
        Err(_) => return violations,
    };

    for entry in entries {
        let message = entry
            .get("eventMessage")
            .and_then(|v| v.as_str())
            .unwrap_or("");

        if !message.contains("deny") {
            continue;
        }

        let timestamp = entry
            .get("timestamp")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_owned();

        // Extract operation: "deny(1) file-read-data" → "file-read-data"
        let operation = extract_operation(message);
        let path = extract_path(message);

        violations.push(SandboxViolation {
            timestamp,
            operation,
            path,
        });
    }

    violations
}

/// Extract the denied operation from a sandbox log message.
/// e.g. "deny(1) file-read-data" → "file-read-data"
fn extract_operation(message: &str) -> String {
    if let Some(pos) = message.find("deny") {
        let after = &message[pos..];
        // Skip "deny" and optional "(N) "
        let rest = after.trim_start_matches("deny");
        let rest = rest.trim_start_matches(|c: char| c == '(' || c.is_ascii_digit() || c == ')');
        let rest = rest.trim_start();
        // Take until whitespace or end
        rest.split_whitespace()
            .next()
            .unwrap_or("unknown")
            .to_owned()
    } else {
        "unknown".to_owned()
    }
}

/// Extract the path from a sandbox log message.
/// Looks for `path "..."` pattern.
fn extract_path(message: &str) -> String {
    if let Some(start) = message.find("path \"") {
        let after = &message[start + 6..];
        if let Some(end) = after.find('"') {
            return after[..end].to_owned();
        }
    }
    String::new()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn extract_operation_basic() {
        assert_eq!(
            extract_operation("deny(1) file-read-data"),
            "file-read-data"
        );
    }

    #[test]
    fn extract_operation_no_parens() {
        assert_eq!(extract_operation("deny file-write-data"), "file-write-data");
    }

    #[test]
    fn extract_path_basic() {
        assert_eq!(
            extract_path("deny(1) file-read-data path \"/usr/local/bin/foo\""),
            "/usr/local/bin/foo"
        );
    }

    #[test]
    fn extract_path_missing() {
        assert_eq!(extract_path("deny(1) network-outbound"), "");
    }

    #[test]
    fn parse_log_output_empty() {
        let violations = parse_log_output("[]");
        assert!(violations.is_empty());
    }

    #[test]
    fn parse_log_output_with_deny() {
        let json = r#"[{"timestamp":"2024-01-01T00:00:00Z","eventMessage":"deny(1) file-read-data path \"/secret\""}]"#;
        let violations = parse_log_output(json);
        assert_eq!(violations.len(), 1);
        assert_eq!(violations[0].operation, "file-read-data");
        assert_eq!(violations[0].path, "/secret");
    }

    #[test]
    fn parse_log_output_skips_non_deny() {
        let json = r#"[{"timestamp":"t1","eventMessage":"allow file-read-data"},{"timestamp":"t2","eventMessage":"deny(1) network-outbound"}]"#;
        let violations = parse_log_output(json);
        assert_eq!(violations.len(), 1);
        assert_eq!(violations[0].operation, "network-outbound");
    }
}
